File: internal/stream/pacer.go
package stream

import (
	"context"
	"time"
)

// Pacer schedules frame emissions on a fixed grid anchored at the first call.
//
// It exists instead of a bare time.Ticker for two reasons:
//
//   - the grid is absolute, so a frame that is delivered slightly late does not
//     permanently shift every later frame (no drift accumulation);
//   - if the sender falls more than one interval behind (a slow client, a GC
//     pause, a saturated link) the grid is resynchronised instead of emitting a
//     catch-up burst that would distort the client's inter-arrival measurements.
type Pacer struct {
	interval time.Duration
	next     time.Time
	started  bool
	timer    *time.Timer
}

// NewPacer returns a Pacer that emits frames every interval.
func NewPacer(interval time.Duration) *Pacer {
	if interval <= 0 {
		interval = time.Millisecond
	}
	return &Pacer{interval: interval, timer: time.NewTimer(0)}
}

// Stop releases the Pacer's timer. The Pacer must not be used afterwards.
func (p *Pacer) Stop() { p.timer.Stop() }

// Wait blocks until the next frame should be emitted. It returns false when the
// deadline has already passed and the loop should finish.
//
// Wait must be called sequentially from a single goroutine.
func (p *Pacer) Wait(ctx context.Context, deadline time.Time) (bool, error) {
	now := time.Now()
	if !p.started {
		p.started = true
		p.next = now
	}

	if d := p.next.Sub(now); d > 0 {
		p.timer.Reset(d)
		select {
		case <-ctx.Done():
			return false, context.Cause(ctx)
		case <-p.timer.C:
		}
		now = time.Now()
	}

	if p.next.Sub(now) < -p.interval {
		// More than one interval behind: resynchronise rather than burst.
		p.next = now
	}

	if !now.Before(deadline) {
		return false, nil
	}

	p.next = p.next.Add(p.interval)
	return true, nil
}

File: dotnet/src/Nsst.Core/Streaming/Pacer.cs
namespace Nsst.Core.Streaming;

/// <summary>
/// Schedules frame emissions on a fixed grid anchored at the first call.
/// </summary>
/// <remarks>
/// It exists instead of a bare timer for two
/// reasons:
/// <list type="bullet">
/// <item>the grid is absolute, so a frame delivered slightly late does not
/// permanently shift every later frame — there is no drift accumulation;</item>
/// <item>if the sender falls more than one interval behind (a slow client, a GC
/// pause, a saturated link) the grid is resynchronised rather than emitting a
/// catch-up burst, which would distort the client's inter-arrival
/// measurements.</item>
/// </list>
/// <para>
/// Timing is taken from <see cref="TimeProvider"/> monotonic timestamps rather
/// than wall clock, so an NTP step during a long test cannot make the grid jump.
/// </para>
/// </remarks>
public sealed class Pacer
{
    private readonly TimeSpan _interval;
    private readonly TimeProvider _timeProvider;
    private readonly IDelayStrategy _delay;
    private readonly long _origin;

    private TimeSpan _next;
    private bool _started;

    /// <summary>Creates a pacer that emits frames every <paramref name="interval"/>.</summary>
    /// <param name="delayStrategy">
    /// The wait primitive. Left null the portable <see cref="TaskDelayStrategy"/> is used,
    /// which is what a <see cref="TimeProvider"/>-driven test clock needs; the server passes
    /// a platform strategy able to reach the sub-tick precision this tool depends on.
    /// </param>
    public Pacer(TimeSpan interval, TimeProvider timeProvider, IDelayStrategy? delayStrategy = null)
    {
        _interval = interval <= TimeSpan.Zero ? TimeSpan.FromMilliseconds(1) : interval;
        _timeProvider = timeProvider;
        _delay = delayStrategy ?? TaskDelayStrategy.Instance;
        _origin = timeProvider.GetTimestamp();
    }

    /// <summary>Monotonic time since this pacer was created.</summary>
    public TimeSpan Now => _timeProvider.GetElapsedTime(_origin, _timeProvider.GetTimestamp());

    /// <summary>
    /// Waits until the next frame should be emitted.
    /// </summary>
    /// <returns>
    /// False when <paramref name="deadline"/> has already passed and the loop
    /// should finish. Throws <see cref="OperationCanceledException"/> when
    /// <paramref name="cancellationToken"/> fires, which the caller classifies.
    /// </returns>
    /// <remarks>Must be called sequentially; a pacer is not safe for concurrent use.</remarks>
    public async ValueTask<bool> WaitAsync(TimeSpan deadline, CancellationToken cancellationToken)
    {
        var now = Now;
        if (!_started)
        {
            _started = true;
            _next = now;
        }

        if (_next > now)
        {
            await _delay.DelayAsync(_next - now, _timeProvider, cancellationToken).ConfigureAwait(false);
            now = Now;
        }

        if (_next - now < -_interval)
        {
            // More than one interval behind: resynchronise rather than burst.
            _next = now;
        }

        if (now >= deadline)
        {
            return false;
        }

        _next += _interval;
        return true;
    }
}
